storage/store.go

```diff
@@ -175,42 +193,27 @@ func (e *NotBootstrappedError) Error() string {
 	return "store has not been bootstrapped"
 }
 
-// storeRangeIterator is an implementation of rangeIterator which
-// cycles through a store's rangesByKey slice.
-type storeRangeIterator struct {
-	store     *Store
-	remaining int
-	index     int
-}
-
-func newStoreRangeIterator(store *Store) *storeRangeIterator {
-	r := &storeRangeIterator{
-		store: store,
-	}
-	r.Reset()
-	return r
+// storeRangeSet is an implementation of rangeSet which
+// cycles through a store's rangesByKey btree with mutex locking.
+type storeRangeSet struct {
+	store *Store
 }
 
-func (si *storeRangeIterator) Next() *Range {
-	si.store.mu.Lock()
-	defer si.store.mu.Unlock()
-	if index, remaining := si.index, len(si.store.rangesByKey)-si.index; remaining > 0 {
-		si.index++
-		si.remaining = remaining - 1
-		return si.store.rangesByKey[index]
-	}
-	return nil
-}
-
-func (si *storeRangeIterator) EstimatedCount() int {
-	return si.remaining
+func (rs *storeRangeSet) Visit(iterator func(*Range) bool) {
+	// Hold the lock except when the iterator is called.
```

Comment:
Replaced `iterator` with `visitor`.

Yes, it's not very clear what is guaranteed.. https://github.com/google/btree/blob/master/btree.go#L393 just says:

```
// Write operations are not safe for concurrent mutation by multiple
// goroutines, but Read operations are.
```

I think `panic` won't happen too as long as we use `store.mu` to protect insertion/deletion/lookup, but yes, there is some risk of skipping ranges.

One naive solution would be just copying ranges to a slice and then iterating over the slice.

```
func (rs *storeRangeSet) Visit(visitor func(*Range) bool) {
	// Hold the lock except when the visitor is called.
	rs.store.mu.RLock()
	ranges := []*Range{}
	rs.store.rangesByKey.Ascend(func(i btree.Item) bool {
		ranges = append(ranges, i.(*Range))
		return true
	})
	rs.store.mu.RUnlock()

	for _, r := range ranges {
		if !visitor(r) {
			break
		}
	}
}
```

We will need extra copy and no new ranges will not be processed until the iteration reaches the end, but it might not be too bad? Need to think more..